### PR TITLE
Add API for marking a syntax tree as generated

### LINKED ISSUE #3705: Exclude files or folder from running through the analyzers through a configuration file.
Hi,

we have a lot of generated files that trigger warnings, errors when analyzers run them.
since the file name of the generated files do not follow the common naming conventions for generated code roslyn fires the analyzers away. As a result there are ten of thousands of warnings in the solution.
Could we get a exclude file from analyzing logic that:
- Exclude a file by adding it to a exclusion list
- Exclude a file by adding it to a exclusion list
- Annotate classes with a special Attribute that exclude the class
- exclude certain file name extensions ( like js files generated by typescript)
- Combination of the above like exclude all js files that live under the folder x including all sub folders

Most of the rules come straight from Resharper's exclusion file manager

Thanks

### PR DESCRIPTION
**Pull request by agocke**
Provides API to fix #3705.

### CODE REVIEW

**Comment by mavasani** [comment]
Tagging @jinujoseph @vatsalyaagrawal @jaredpar 

Is this PR still targeted for 16.5? If not, can we please merge the branch https://github.com/dotnet/roslyn/tree/features/editorconfig-generated-code back into master for 16.5, so the core feature requests in #3705 are unblocked (more context [here](https://github.com/dotnet/roslyn/issues/3705#issuecomment-537225988))? I understand compiler team's desire to ship both analyzer and compiler features in same release, but given there isn't a consensus on this API change in the compiler team, it does not seem reasonable to block a highly requested analyzer feature on this PR.

File: src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.cs [hunk 1]
```diff
@@ -308,7 +308,8 @@ private void BuildPreprocessorStateChangeMap()
             CSharpParseOptions options = null,
             string path = "",
             Encoding encoding = null,
-            ImmutableDictionary<string, ReportDiagnostic> diagnosticOptions = null)
+            ImmutableDictionary<string, ReportDiagnostic> diagnosticOptions = null,
+            bool? isGeneratedCode = null)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.cs]: This is a public API, we need to add backcompat overloads.

File: src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.cs [hunk 2]
```diff
@@ -374,9 +385,10 @@ internal static SyntaxTree CreateWithoutClone(CSharpSyntaxNode root)
             string path = "",
             Encoding encoding = null,
             ImmutableDictionary<string, ReportDiagnostic> diagnosticOptions = null,
+            bool? isGeneratedCode = null,
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.cs]: Please add a backcompat overload

File: src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.cs [hunk 3]
```diff
@@ -387,6 +399,7 @@ internal static SyntaxTree CreateWithoutClone(CSharpSyntaxNode root)
             CSharpParseOptions options = null,
             string path = "",
             ImmutableDictionary<string, ReportDiagnostic> diagnosticOptions = null,
+            bool? isGeneratedCode = null,
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.cs]: Please add a backcompat overload

File: src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.cs (hunk 2 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.cs]: Ah, I see you added these below. Sorry for the spam

---
In reply to: [346550550](https://github.com/dotnet/roslyn/pull/39051#discussion_r346550550) [](ancestors = 346550550)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 4]
```diff
@@ -5597,6 +5597,65 @@ static void H()
...
+            var source3 = CSharpSyntaxTree.ParseText(
+@"
+partial class Program
+{
+    static void I(object x)
+    {
+        x = null; // warning 3
+        F = null; // warning 4
+    }
+}", isGeneratedCode: null /* use heuristic */);
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >/* use heuristic */ [](start = 26, length = 19)

I'm not really sure what this comment is supposed to indicate. Consider expanding or removing.

File: src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.cs [hunk 5]
```diff
@@ -319,6 +320,13 @@ private void BuildPreprocessorStateChangeMap()
                 ((CompilationUnitSyntax)root).GetConditionalDirectivesStack() :
                 InternalSyntax.DirectiveStack.Empty;
 
+            bool isGenerated = isGeneratedCode.HasValue
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxTree.cs]: >isGeneratedCode [](start = 31, length = 15)

bool isGenerated = isGeneratedCode ?? GeneratedCodeUtilities...

File: src/Compilers/Core/Portable/InternalUtilities/GeneratedCodeUtilities.cs [hunk 6]
```diff
@@ -76,6 +82,35 @@ private static bool IsGeneratedCodeFile(string filePath)
             return false;
         }
 
+        private static bool BeginsWithAutoGeneratedComment(SyntaxNode root, Func<SyntaxTrivia, bool> isComment)
```
Comment (mavasani) [inline src/Compilers/Core/Portable/InternalUtilities/GeneratedCodeUtilities.cs]: Sorry for the delayed comment, but seems like you can share majority of code between this method and the overload below which takes `SyntaxTree` argument.

**Pull request merged**